File: LXDSlideMenu-master/LXDSlideMenuObjectiveC/LXDSlideMenuObjectiveC/LXDSlideManager.m
//
//  LXDSlideManager.m
//  LXDSlideMenuObjectiveC
//
//  Created by linxinda on 15/11/29.
//  Copyright © 2015年 sindriLin. All rights reserved.
//

#import "LXDSlideManager.h"


#define kMaxOpenRatio 0.7
#define kWidth CGRectGetWidth([UIScreen mainScreen].bounds)


@interface LXDSlideManager ()

@property (nonatomic, strong) UITapGestureRecognizer * tap;
@property (nonatomic, strong) UIView * snapView;
@property (nonatomic, assign) BOOL isOpen;

@end


@implementation LXDSlideManager

static LXDSlideManager * sharedManager = nil;
UIViewController * _mainController = nil;
UIViewController * _menuController = nil;

+ (__nonnull instancetype)sharedManager
{
    static dispatch_once_t once;
    dispatch_once(&once, ^{
        sharedManager = [[self alloc] init];
    });
    return sharedManager;
}


- (void)setupMainController: (UIViewController *)mainController
{
    [_mainController.view removeGestureRecognizer: self.tap];
    [_mainController.view removeFromSuperview];
    _mainController = mainController;
    
    if (_menuController == nil) { return; }
    [_mainController.view addSubview: _menuController.view];
}

- (void)setupMenuController:(UIViewController *)menuController
{
    [_menuController.view removeFromSuperview];
    _menuController = menuController;
    _menuController.view.frame = CGRectOffset([UIScreen mainScreen].bounds, kWidth, 0);
    
    if (_demoType == LXDDemoTypeOnWindow) {
        [[UIApplication sharedApplication].keyWindow addSubview: _menuController.view];
    } else if (_demoType == LXDDemoTypeOnMainView && _mainController != nil) {
        [_mainController.view addSubview: _menuController.view];
    }
}

- (void)dealloc
{
    _mainController = nil;
    _menuController = nil;
}

- (UITapGestureRecognizer *)tap
{
    if (!_tap) {
        _tap = [[UITapGestureRecognizer alloc] initWithTarget: self action: @selector(close)];
    }
    return _tap;
}


- (void)setDemoType:(LXDDemoType)demoType
{
    _demoType = demoType;
    [_menuController.view removeFromSuperview];
    if (_menuController == nil) { return; }
    
    if (demoType == LXDDemoTypeOnWindow) {
        [[UIApplication sharedApplication].keyWindow addSubview: _menuController.view];
    } else if (demoType == LXDDemoTypeOnMainView && _mainController != nil) {
        [_mainController.view addSubview: _menuController.view];
    }
}


- (void)openOrClose
{
    if (_isOpen) {
        [self close];
    } else {
        [self open];
    }
}

- (BOOL)useSnap {
    return _slideType == LXDSlideTypeBothMove && _demoType == LXDDemoTypeOnMainView;
}


- (void)open
{
    _isOpen = YES;
    
    if (_menuController == nil || _mainController == nil) {
        @throw [NSException exceptionWithName: @"LXDNullInstanceException" reason: @"you can not slide view when one or more of menu view and main view is nil" userInfo: nil];
    }
    
    CGRect mainFrame = _mainController.view.frame;
    CGRect menuFrame = _menuController.view.frame;
    
    if (_slideType == LXDSlideTypeMenuMove || _demoType == LXDDemoTypeOnWindow) {
        menuFrame.origin.x = (1 - kMaxOpenRatio) * kWidth;
    }
    if (_slideType == LXDSlideTypeBothMove) {
        mainFrame.origin.x = -kWidth * kMaxOpenRatio;
    }
    if ([self useSnap]) {
        self.snapView = [_mainController.view snapshotViewAfterScreenUpdates: NO];
        _snapView.frame = _mainController.view.frame;
        [_mainController.view addSubview: _snapView];
        menuFrame.origin.x = (1 - kMaxOpenRatio) * kWidth;
    }
    
    [UIView animateWithDuration: 0.2 animations: ^{
        if ([self useSnap]) {
            _snapView.frame = mainFrame;
        } else {
            _mainController.view.frame = mainFrame;
        }
        _menuController.view.frame = menuFrame;
    } completion: ^(BOOL finished) {
        if (finished) {
            [self useSnap] ? [_snapView addGestureRecognizer: self.tap] : [_mainController.view addGestureRecognizer: self.tap];
        }
    }];
}

- (void)close
{
    _isOpen = NO;
    [_menuController.view endEditing: YES];
    [_mainController.view removeGestureRecognizer: self.tap];
    
    CGRect menuFrame = _menuController.view.frame;
    CGRect mainFrame = _mainController.view.frame;
    
    mainFrame.origin.x = 0;
    if (_slideType == LXDSlideTypeMenuMove || _demoType == LXDDemoTypeOnWindow || [self useSnap]) {
        menuFrame.origin.x = kWidth;
    }
    
    [UIView animateWithDuration: 0.2 animations: ^{
        if ([self useSnap]) { _snapView.frame = mainFrame; }
        _mainController.view.frame = mainFrame;
        _menuController.view.frame = menuFrame;
    } completion: ^(BOOL finished) {
        [_snapView removeFromSuperview];
    }];
}




@end

File: LXDSlideMenu-master/LXDSlideMenuSwift/LXDSlideMenuSwift/LXDSlideManager.swift
//
//  LXDSlideManager.swift
//  LXDSlideMenuSwift
//
//  Created by linxinda on 15/11/29.
//  Copyright © 2015年 sindriLin. All rights reserved.
//

import UIKit

public enum LXDSlideMenuType: Int {
    case BothMove           //主界面跟侧滑界面共同移动
    case MenuMove          //只移动侧滑菜单界面
}

public enum LXDDemoType: Int {
    case OnWindow          //加在keyWindow上，如果不是从AppDelegate进行rootViewController的初始化，此选项无效
    case OnMainView       //加在主视图上
}


let maxOpenRatio: CGFloat = 0.7             //最大滑动屏幕宽度比例
let kWidth: CGFloat = UIScreen.mainScreen().bounds.size.width

var kMainController: UIViewController?  //主视图控制器
var kMenuController: UIViewController? //边栏视图控制器

class LXDSlideManager: NSObject {
    
    private var isOpen: Bool?
    var slideType: LXDSlideMenuType! = .BothMove
    private var demoType: LXDDemoType! = .OnMainView
    private var snapView: UIView?
    
    lazy private var tap: UITapGestureRecognizer? = {
        return UITapGestureRecognizer(target: self, action: "openOrClose")
    }()
    
    
    /** 单例*/
    private static let let_sharedManager: LXDSlideManager = LXDSlideManager()
    class func sharedManager() -> LXDSlideManager {
        return let_sharedManager
    }
    
    /** 让构造器私有化无法使用*/
    private override init() {
        super.init()
    }
    
    deinit {
        kMainController = nil
        kMenuController = nil
    }
}


extension LXDSlideManager {
    
    func setMainController(mainController: UIViewController!) {
        kMainController?.view.removeGestureRecognizer(self.tap!)
        kMainController?.view.removeFromSuperview()
        kMainController = mainController
        
        if kMenuController != nil {
            kMenuController?.view.removeFromSuperview()
            
            if demoType == .OnMainView {
                mainController.view.addSubview((kMenuController?.view)!)
            }
        }
    }
    
    func setMenuController(menuController: UIViewController!) {
        kMenuController?.view.removeFromSuperview()
        kMenuController = menuController
        let menuFrame: CGRect = CGRectOffset(UIScreen.mainScreen().bounds, kWidth, 0)
        menuController.view.frame = menuFrame
        
        if kMainController != nil && demoType == .OnMainView {
            kMainController?.view.addSubview((menuController.view)!)
        } else if demoType == .OnWindow {
            AppDelegate.getDelegate().window?.addSubview(menuController.view)
        }
    }
    
    func setDemoType(demoType: LXDDemoType) {
        self.demoType = demoType
        kMenuController?.view.removeFromSuperview()
        
        if demoType == .OnWindow {
            if kMenuController != nil {
                AppDelegate.getDelegate().window?.addSubview((kMenuController?.view)!)
            }
        } else if demoType == .OnMainView {
            if kMainController != nil && kMenuController != nil {
                kMainController?.view.addSubview((kMenuController?.view)!)
            }
        }
    }
}


extension LXDSlideManager {
    
    func openOrClose() {
        if isOpen == true {
            self.close()
        } else {
            self.open()
        }
    }
    
    private func useSnap() -> Bool {
        return slideType == .BothMove && demoType == .OnMainView
    }
    
    private func open() {
        isOpen = true
        
        if kMenuController == nil || kMainController == nil {
            fatalError("you can not slide view when one or more of menu view and main view is nil")
        }
        
        var menuFrame: CGRect = (kMenuController?.view.frame)!
        var mainFrame: CGRect = (kMainController?.view.frame)!
        snapView = (kMainController?.view.snapshotViewAfterScreenUpdates(false))!
        
        if slideType == .MenuMove || self.demoType == .OnWindow {
            menuFrame.origin.x = (1 - maxOpenRatio) * kWidth
        }
        if slideType == .BothMove {
            if self.useSnap() {
                snapView!.frame = (kMainController?.view.frame)!
                kMainController?.view.addSubview(snapView!)
                menuFrame.origin.x = (1 - maxOpenRatio) * kWidth
            }
            mainFrame.origin.x = -kWidth * maxOpenRatio
        }
        UIView.animateWithDuration(0.2, animations: { () -> Void in
            if self.useSnap() {
                self.snapView?.frame = mainFrame
            } else {
                kMainController?.view.frame = mainFrame
            }
            kMenuController?.view.frame = menuFrame
            },  completion: { (finished: Bool) -> Void in
                if finished == true {
                    if self.useSnap() {
                        self.snapView?.addGestureRecognizer(self.tap!)
                    } else {
                        kMainController?.view.addGestureRecognizer(self.tap!)
                    }
                }
        })
    }
    
    private func close() {
        isOpen = false
        kMenuController?.view.endEditing(true)
        kMainController?.view.removeGestureRecognizer(self.tap!)
        var menuFrame: CGRect = (kMenuController?.view.frame)!
        var mainFrame: CGRect = (kMainController?.view.frame)!
        mainFrame.origin.x = 0
        
        if slideType == .MenuMove || demoType == .OnWindow {
            menuFrame.origin.x = kWidth
        }
        UIView.animateWithDuration(0.2, animations: { () -> Void in
            if self.useSnap() {
                self.snapView?.frame = mainFrame
                menuFrame.origin.x = kWidth
                kMenuController?.view.frame = menuFrame
            } else {
                kMainController?.view.frame = mainFrame
            }
            kMenuController?.view.frame = menuFrame
            }, completion:  { (finished: Bool) -> Void in
                if finished {
                    self.snapView?.removeFromSuperview()
                }
        })
    }
}

